src/data/sym2.lean

```diff
@@ -0,0 +1,364 @@
+begin
+  split,
+  { intros h z hr hd,
+    induction z,
+    have hd' := (is_diag_iff_proj_eq _).mp hd,
+    have hr' := from_rel_proj_prop.mp hr,
+    rw hd' at hr',
+    exact h _ hr',
+    refl, },
+  { intros h x hr,
+    rw ← @from_rel_prop _ _ sym at hr,
+    exact h hr ⟨x, rfl⟩, }
+end
+
+end relations
+
+section sym_equiv
+
+/-! ### Equivalence to the second symmetric power -/
+
+local attribute [instance] vector.perm.is_setoid
+
+private def from_vector {α : Type*} : vector α 2 → α × α
+| ⟨[a, b], h⟩ := (a, b)
+
+private lemma perm_card_two_iff {α : Type*} {a₁ b₁ a₂ b₂ : α} :
+  [a₁, b₁].perm [a₂, b₂] ↔ (a₁ = a₂ ∧ b₁ = b₂) ∨ (a₁ = b₂ ∧ b₁ = a₂) :=
+begin
+  split, swap,
+  intro h, cases h; rw [h.1, h.2], apply list.perm.swap', refl,
+  intro h,
+  rw ←multiset.coe_eq_coe at h,
+  repeat {rw ←multiset.cons_coe at h},
+  repeat {rw multiset.cons_eq_cons at h},
+  cases h,
+  { cases h, cases h_right,
+    left, simp [h_left, h_right.1],
+    simp at h_right, tauto, },
+  { rcases h.2 with ⟨cs, h'⟩,
+    repeat {rw multiset.cons_eq_cons at h'},
+    simp only [multiset.zero_ne_cons, multiset.coe_nil_eq_zero, exists_false, or_false, and_false, false_and] at h',
+    right, simp [h'.1.1, h'.2.1], },
+end
+
+/--
+The symmetric square is equivalent to length-2 vectors up to permutations.
+-/
+def sym2_equiv_sym' {α : Type*} : equiv (sym2 α) (sym' α 2) :=
+{ to_fun := quotient.map (λ (x : α × α), ⟨[x.1, x.2], rfl⟩) (begin
+    intros x y h, cases h, refl, apply list.perm.swap', refl,
+  end),
+  inv_fun := quotient.map from_vector (begin
+    rintros ⟨x, hx⟩ ⟨y, hy⟩ h,
+    cases x with x0 x, simp at hx, tauto,
+    cases x with x1 x, simp at hx, exfalso, linarith [hx],
+    cases x with x2 x, swap, simp at hx, exfalso, linarith [hx],
+    cases y with y0 y, simp at hy, tauto,
+    cases y with y1 y, simp at hy, exfalso, linarith [hy],
+    cases y with y2 y, swap, simp at hy, exfalso, linarith [hy],
+    cases perm_card_two_iff.mp h; rw [h_1.1, h_1.2],
+    refl,
+    apply sym2.rel.swap,
+  end),
+  left_inv := by tidy,
+  right_inv := begin
+    intro x,
+    induction x,
+    cases x with x hx,
+    cases x with x0 x, simp at hx, tauto,
+    cases x with x1 x, simp at hx, exfalso, linarith [hx],
+    cases x with x2 x, swap, simp at hx, exfalso, linarith [hx],
+    refl,
+    refl,
+  end }
+
+/--
+The symmetric square is equivalent to the second symmetric power.
+-/
+def equiv_sym (α : Type*) : sym2 α ≃ sym α 2 :=
+equiv.trans sym2_equiv_sym' (sym_equiv_sym' α 2).symm
+
+/--
+The symmetric square is equivalent to multisets of cardinality
+two. (This is currently a synonym for `equiv_sym`, but it's provided
+in case the definition for `sym` changes.)
+-/
+def equiv_multiset (α : Type*) : sym2 α ≃ {s : multiset α // s.card = 2} :=
+equiv_sym α
+
+end sym_equiv
+
+section inhabited
+
+-- Instances to make the linter happy
+
+instance inhabited_vmem [inhabited α] : inhabited (vmem (default α) (diag (default α))) :=
```

Comment:
Oh sorry, I didn't look carefully why the linter was complaining, I thought it was about the definition `sym2` itself (it isn't complaining about `sym2` because you marked it as reducible).

This instance is useless, and the linter shouldn't complain about `vmem`. Please remove this instance and add `@[nolint has_inhabited_instance]` in front of the definition of `vmem`.